Pasfmt.Cursors.pas

```diff
@@ -0,0 +1,339 @@
+  Inc(Copied, Length(FOffsets));
+
+  for List in [FBookmarkOffsets, FBreakpointOffsets] do begin
+    if List.Count > 0 then
+      Move(List.List[0], Result[Copied], List.Count * SizeOf(List[0]));
+    Inc(Copied, List.Count);
+  end;
+end;
+
+//______________________________________________________________________________________________________________________
+
+procedure TCursors.Deserialize(Offsets: TArray<Integer>);
+var
+  ExpectedCount: Integer;
+  Copied: Integer;
+  List: TList<Integer>;
+begin
+  ExpectedCount := Length(FOffsets) + FBookmarkOffsets.Count + FBreakpointOffsets.Count;
+  if Length(Offsets) <> ExpectedCount then begin
+    Log.Error(
+        'Expected %d cursors, found %d. Editor cursors, breakpoints, and bookmarks cannot be updated',
+        [ExpectedCount, Length(Offsets)]
+    );
+    Exit;
+  end;
+
+  Copied := 0;
+
+  if Length(FOffsets) > 0 then
+    Move(Offsets[Copied], FOffsets[0], Length(FOffsets) * SizeOf(FOffsets[0]));
+  Inc(Copied, Length(FOffsets));
+
+  for List in [FBookmarkOffsets, FBreakpointOffsets] do begin
+    if List.Count > 0 then
+      Move(Offsets[Copied], List.List[0], List.Count * SizeOf(List[0]));
+    Inc(Copied, List.Count);
+  end;
+end;
+
+//______________________________________________________________________________________________________________________
+
+procedure TCursors.UpdateBookmarks(Buffer: IOTAEditBuffer);
+var
+  EditView: IOTAEditView;
+  EditPos: TOTAEditPos;
+  CharPos: TOTACharPos;
+
+  BookmarkIdx: Integer;
+begin
+  if Buffer.EditViewCount <= 0 then
+    Exit;
+
+  // bookmarks are shared between all edit views of a buffer
+  EditView := Buffer.EditViews[0];
+
+  for BookmarkIdx := 0 to FBookmarkIds.Count - 1 do begin
+    CharPos := EditView.PosToCharPos(FBookmarkOffsets[BookmarkIdx]);
+    EditView.ConvertPos(False, EditPos, CharPos);
+    EditView.CursorPos := EditPos;
+
+    EditView.BookmarkToggle(FBookmarkIds[BookmarkIdx]);
+  end;
+end;
+
+//______________________________________________________________________________________________________________________
+
+procedure TCursors.UpdateBreakpoints(Buffer: IOTAEditBuffer);
+var
+  I: Integer;
+  CharPos: TOTACharPos;
+  BreakpointData: ^TBreakpointData;
+  Breakpoint: IOTABreakpoint;
+  DebuggerServices: IOTADebuggerServices;
+begin
+  if BorlandIDEServices.QueryInterface(IOTADebuggerServices, DebuggerServices) <> S_OK then begin
+    Log.Debug('Skipping updating breakpoints because BorlandIDEServices does not implement IOTADebuggerServices');
+    Exit;
+  end;
+
+  // The old breakpoints (sometimes) stick around after a rewrite of the editor, but their positions are wrong and
+  // they become invisible until manually disabled and re-enabled, so it's better to just recreate them.
+  for I := DebuggerServices.SourceBkptCount - 1 downto 0 do begin
+    Breakpoint := DebuggerServices.SourceBkpts[I];
+    if Breakpoint.FileName <> Buffer.FileName then
```

Comment:
Might be a good idea to use a `SameText` here in case the file casing is different between the two.